### PR TITLE
storage: add support for AWS IAM authentication for Kafka

### PR DESCRIPTION
**Pull request by benesch**
This has been a long desired feature.

Work towards a fix for https://github.com/MaterializeInc/materialize/issues/8237.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

First commit is all code movement. Focus on the second commit.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add support for IAM authentication when creating Kafka connections to AWS MSK clusters.

### CODE REVIEW

File: src/kafka-util/src/aws.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+) -> Result<(String, i64), SignerError> {
+    let credentials = credentials_provider.provide_credentials().await?;
+    construct_auth_token(region, credentials)
+}
+
+fn construct_auth_token(
+    region: Region,
+    credentials: Credentials,
+) -> Result<(String, i64), SignerError> {
+    let endpoint_url = format!("https://kafka.{}.amazonaws.com", region);
```
Comment (rjobanp) [inline src/kafka-util/src/aws.rs:64]: shouldn't this be configured based on the endpoint_url of the AWS Connection object? I don't know if any other providers follow the MSK API but we do allow customers to set the endpoint here https://materialize.com/docs/sql/create-connection/#aws-options and it seems unusual to ignore it for this auth check

**Review by rjobanp — APPROVED** [review]
nice work!

code LGTM - just one question on configurability of the MSK auth endpoint

File: src/kafka-util/src/aws.rs (hunk 1 above)
Comment (jubrad) [reply src/kafka-util/src/aws.rs:64]: There's also some non standard endpoints in AWS for FIPs I think predominantly used in gov cloud. I don't think we need to seriously consider supporting these, but it's worth calling out the limitations of this endpoint construction. 
https://docs.aws.amazon.com/general/latest/gr/rande.html#FIPS-endpoints

File: doc/user/content/sql/create-connection.md [hunk 2]
```diff
@@ -307,6 +308,23 @@ CREATE CONNECTION kafka_connection TO KAFKA (
...
 ```
 {{< /tab >}}
+
+{{< tab "AWS IAM">}}
+
+{{< warn-if-unreleased "v0.112" >}}
+
+```mzsql
+CREATE CONNECTION aws_msk TO AWS (
+    ASSUME ROLE ARN = 'arn:aws:iam::400121260767:role/MaterializeMSK',
```
Comment (bobbyiliev) [inline doc/user/content/sql/create-connection.md:318]:
```suggestion
    ASSUME ROLE ARN = 'arn:aws:iam::400121260767:role/MaterializeMSK'
 ```

File: src/kafka-util/src/client.rs [hunk 3]
```diff
@@ -438,7 +444,30 @@ impl<C> ClientContext for TunnelingClientContext<C>
 where
     C: ClientContext,
 {
-    const ENABLE_REFRESH_OAUTH_TOKEN: bool = C::ENABLE_REFRESH_OAUTH_TOKEN;
+    const ENABLE_REFRESH_OAUTH_TOKEN: bool = true;
+
+    fn generate_oauth_token(
```
Comment (jubrad) [inline src/kafka-util/src/client.rs:449]: Is it worth a comment here indicating that the only oath we support will be for AWS?

**Review by jubrad — APPROVED** [review]
Very Nice!

File: src/kafka-util/src/aws.rs (hunk 1 above)
Comment (benesch) [reply src/kafka-util/src/aws.rs:68]: Good point! I do think in theory we want to pull this from the AWS configuration in the connection, but the AWS SDK is inscrutable on how to do this. I dropped a comment in but I'm just going to leave this as is for now. AFAIK nothing besides real AWS supports MSK IAM auth, so there's no way to test this with localstack or anything ([localstack ticket](https://github.com/localstack/localstack/issues/10617)), and so the hardcoding here is ~fine. (Yeah, doesn't support FIPS, but that's a problem for a long time for now.)

File: doc/user/content/sql/create-connection.md (hunk 2 above)
Comment (benesch) [reply doc/user/content/sql/create-connection.md]: Oops, thanks, fixed!

File: doc/user/content/sql/create-connection.md [hunk 4]
```diff
@@ -307,6 +308,23 @@ CREATE CONNECTION kafka_connection TO KAFKA (
...
+);
+
+CREATE CONNECTION kafka_msk TO KAFKA (
+    BROKER 'msk.mycorp.com:9092',
+    SECURITY PROTOCOL = 'SASL_SSL',
+    AWS CONNECTION = kafka_msk
```
[the remark this answers is not in the document]
Comment (benesch) [reply doc/user/content/sql/create-connection.md]: Fixed as well.

File: src/kafka-util/src/client.rs (hunk 3 above)
Comment (benesch) [reply src/kafka-util/src/client.rs:449]: Sure! Done.

**Comment by benesch** [comment]
Got this passing CI, and seems to work in Bobby's staging environment too. Merging to avoid bitrot, but I tossed in a feature flag so we can do some more testing in our internal environments before we send out to the world. Thanks for the reviews!

**Pull request merged**